app/services/policies/policies_list_service.rb

```diff
@@ -4,16 +4,17 @@
 require 'set'
 
 class Policies::PoliciesListService
+
+  class PoliciesListServiceError < RuntimeError; end
+
   def self.call(account, builtin: true)
     list = PolicyList.new
-    if builtin
-      response = ::JSONClient.get(ThreeScale.config.sandbox_proxy.apicast_registry_url)
-      return unless response.ok?
-
-      list.merge!(PolicyList.from_hash(response.body['policies']))
-    end
+    merge_policies_from_apicast(list) if builtin
```

Comment:
What about something like this?
```ruby
def self.call(account, builtin: true)
    hashes = []
    hashes << fetch_policies_from_apicast if builtin
    hashes << account.policies if account.provider_can_use?(:policy_registry)
    PolicyList.from_hash(*hashes)
rescue PoliciesListServiceError => error
    Rails.logger.error { error } and return
end
```

And the method `from_hash` changed to:
```ruby
def self.from_hash(*hashes)
  object = new
  hashes.each { |hash| object.merge!(hash.as_json) }
  object
end
```

Reply:
Looks good!

Reply:
right! I will write you back with a simple update, one moment 😄

Reply:
```ruby
# frozen_string_literal: true

require 'jsonclient'
require 'set'

class Policies::PoliciesListService

  class PoliciesListServiceError < RuntimeError; end

  def self.call(account, builtin: true)
    lists = []
    lists << PolicyList.from_hash(fetch_policies_from_apicast) if builtin
    lists << PolicyList.new(account.policies) if account.provider_can_use?(:policy_registry)
    PolicyList.merge_policy_lists(*lists).to_h
  rescue PoliciesListServiceError => error
    Rails.logger.error { error } and return
  end

  class PolicyList
    include Enumerable

    attr_reader :sets
    protected :sets
    delegate :each, to: :sets

    def initialize(policies = [])
      @sets = Hash.new { |hash, key| hash[key] = Set.new }
      policies.each(&method(:add))
    end

    # This smells :reek:FeatureEnvy
    # but it is OK
    def add(policy)
      @sets[policy.name.to_s].add(policy.schema)
    end

    def merge(other)
      object = dup
      object.merge!(other)
      object
    end

    def merge!(other)
      @sets.deep_merge!(other.to_h) do |_key, values, other_values|
        values + other_values
      end
    end

    def initialize_copy(source)
      super
      @sets = source.sets.dup
    end

    def self.from_hash(hash)
      object = new
      object.merge!(hash.as_json)
      object
    end

    def self.merge_policy_lists(*policy_lists)
      object = new
      policy_lists.each { |policy_list| object.merge!(policy_list) }
      object
    end

    def to_h
      @sets.transform_values(&:to_a).as_json
    end
  end

  HTTP_ERRORS = [HTTPClient::BadResponseError, HTTPClient::TimeoutError, HTTPClient::ConnectTimeoutError, HTTPClient::SendTimeoutError, HTTPClient::ReceiveTimeoutError, SocketError].freeze
  private_constant :HTTP_ERRORS

  def self.fetch_policies_from_apicast
    apicast_registry_url = ThreeScale.config.sandbox_proxy.apicast_registry_url
    begin
      response = ::JSONClient.get(apicast_registry_url)
      return response.body['policies'] if response.ok?
      raise PoliciesListServiceError, "APIcast could not be found at url: '#{apicast_registry_url}'. Error: #{response.content}"
    rescue *HTTP_ERRORS => error
      raise PoliciesListServiceError, "APIcast could not be found at url: '#{apicast_registry_url}'. Error: #{error}"
    end
  end

  private_class_method :fetch_policies_from_apicast
end
```

Reply:
the changes are only 2:
```ruby
def self.call(account, builtin: true)
    lists = []
    lists << PolicyList.from_hash(fetch_policies_from_apicast) if builtin
    lists << PolicyList.new(account.policies) if account.provider_can_use?(:policy_registry)
    PolicyList.merge_policy_lists(*lists).to_h
  rescue PoliciesListServiceError => error
    Rails.logger.error { error } and return
end

# ...

 def self.merge_policy_lists(*policy_lists)
      object = new
      policy_lists.each { |policy_list| object.merge!(policy_list) }
      object
end
```

Reply:
I did my own implementation in the end, I think there were some unnecessary steps (like too many to_hash / to_json for nothing). Codeclimate is still complaining tho!